BLW-34 Text-Alerts

felix-berlin opened the pull request:
Dieser PR bringt eine neue Alert Variation `.alert--text-xyz`.

Unter http://blower-testground.test/components/alert.php gibt es Beispiele.

Die neue sowie ältere Varianten werden nun mittels Mixin und Loop generiert.

felix-berlin committed:
feat(BLW-34): add alert--text modifier

felix-berlin committed:
test(BLW-34): add tests for alert

kaiserkiwi commented on dist/css/blower.css:
```diff
@@ -370,6 +370,24 @@
+      --border-width-alert: 0;
+      display: inline-block;
+    }
+    .alert--text-success {
+      --color-alert-text: #3F9A7A;
+    }
+    .alert--text-danger {
+      --color-alert-text: #D75052;
+    }
+    .alert--text-warning {
+      --color-alert-text: #DF8260;
+    }
+    .alert--text-info {
+      --color-alert-text: #00B8D4;
+    }
```
Bin ich kein Freund von. Das geht ein bisschen gegen BEM.

Mein Vorschlag wäre, dass man hier `.alert--text` und `.alert--${status}` kombiniert und dann die entsprechenden Custom Properties überschreibt.

Schau dir gerne mal an, wie das bei den Button-Klassen gelöst ist. Hier wird z.B. ein destruktiver Outline-Button wie folgt definiert:

```html
<button class="button button--destructive button--outline">
	Destructive
</button>
```

Die Klassen sind also `button`, `button--outline` und `button--destructive` und nicht `button`, `button--outline-destructive`.

--- 

Ich würde an der Stelle mal zur Debatte stellen, ob wir nicht vom BEM-Standard hier etwas abweichen wollen und stattdessen zu `is-text` und `is-success` wechseln. Dann ist das Konsistent zwischen allen Komponenten. Könnte aber eben auch verwirrend sein, weil `.button.is-text` nicht dieselben Styles anwenden, wie `.alert.is-text`, sondern eben je nach Komponente unterschieden wird. An sich ein Teil von BEMIT.

Das wäre allerdings ein Breaking Change, der nicht in diesem Pull Request bearbeitet werden sollte, sondern in einem separaten Ticket.

felix-berlin replied:
Ich glaube, ich verstehe, was du meinst, hatte es mit diesem Ansatz versucht, bin allerdings nicht zu dem gewünschten Ergebnis gekommen. Meine Erkenntnis war, es muss viel überschrieben werde und die Farben können leider nicht aus den vorhandenen vars recycelt werden (da schlicht nicht vorhanden). So, hab ich mich dann für diese Variante entschieden. Also in irgendeiner Weise müssen die Text-Varianten ihre eigene Farbe mitbringen, sonst ginge das wahrscheinlich ganz ok. Oder `.alert--${status}` speichert diese bereits, aber das ist auch etwas komisch, oder?

kaiserkiwi replied:
Ok, ich hab mir mal Zeit genommen, um mir das anzuschauen und es entsprechend angepasst. Nun können wir problemlos mit `class="alert alert--success alert--text"` arbeiten.

kaiserkiwi commented on scss/components/_alert.scss:
```diff
@@ -3,6 +3,26 @@
+	'danger': (color.$danger-200, color.$danger-900, color.$danger-300),
+	'warning': (color.$warning-200, color.$warning-900, color.$warning-300),
+	'info': (color.$gray-200, color.$gray-900, color.$gray-300),
+	'dark': (color.$gray-600, #FFF, color.$gray-900),
+	'light': (color.$gray-100, color.$gray-900, color.$gray-200)
```
Diese werden nie verwendet, richtig?

felix-berlin replied:
Die beiden werden wie die anderen durch diesen Loop verarbeitet.

```scss
	@each $type, $colors in $alert-types {
		@include mx.alert-type($type, map-get($colors, 'background'), map-get($colors, 'text'), map-get($colors, 'border-color'));
	}
```

kaiserkiwi commented on scss/components/_alert.scss:
```diff
@@ -3,6 +3,26 @@
+/// Sorted by background, text and border color
+/// @type Map
+$alert-types: (
+	'success': (color.$positive-200, color.$positive-900, color.$positive-300),
+	'danger': (color.$danger-200, color.$danger-900, color.$danger-300),
+	'warning': (color.$warning-200, color.$warning-900, color.$warning-300),
+	'info': (color.$gray-200, color.$gray-900, color.$gray-300),
```
Ich bin mir nicht sicher, ob ich das so gut finde. Kann man hier keine Named Parameter übergeben?

Also stattdessen:
```sass
'success': (
	background: color.$positive-200, 
	text: color.$positive-900, 
	border: color.$positive-300
),
```

Ich hab gerade nicht auf dem Schirm ob Sass das kann, aber es wäre um ein Vielfaches lesbarer.

felix-berlin replied:
Da haste Recht, hab die Maps entsprechend angepasst.

felix-berlin committed:
refactor(BLW-34): use key value map

felix-berlin committed:
style(BLW-34): reformat

felix-berlin committed:
Merge remote-tracking branch 'origin/blw-33-sass-tests-schreiben' into blw-34-text-alerts

# Conflicts:
#	dist/css/blower.min.css
#	package-lock.json

kaiserkiwi committed:
refactor: Change the way the text alerts are written
- Instead of having complete new classes, the `.alert--text` modifier now allows to make every alert to a text alert
- Reduce specificity with `:where()`
- Add dark and light to text alerts

kaiserkiwi committed:
Merge branch 'main' into blw-34-text-alerts

kaiserkiwi commented:
@felix-berlin Bitte die Tests noch fixen.

felix-berlin committed:
test(BLW-34): fix tests

felix-berlin committed:
chore(BLW-34): commit build output

felix-berlin committed:
chore(BLW-34): commit prod output

felix-berlin committed:
Merge remote-tracking branch 'origin/main' into blw-34-text-alerts

# Conflicts:
#	dist/css/blower.min.css

felix-berlin commented:
@kaiserkiwi Test sind jetzt auch endlich gefixt

felix-berlin committed:
test(BLW-34): fix px to rem function

felix-berlin committed:
build(BLW-34): update true

felix-berlin commented:
Nach dem Merge-Konflikt gibt leider weitere Tests, die fehlschlagen. Ich bin dran.

felix-berlin committed:
fix(BLW-34): add missing error handling in fluid()

felix-berlin commented:
So, jetzt aber :)
